### ISSUE TITLE
data-uri issue quote issue

### ISSUE THREAD

**Issue by leggomuhgreggo** [jakubpawlowicz/clean-css#676]
Using clean-css through gulp-minify-css

Here are the specified settings:
`.pipe(minifyCSS({keepBreaks: false, processImport: false})); `

A use case like this:
`.icon::before{
  background-image: url(data:image/png;base64,iVB...
}`
will compile to:
`.icon::before{
  background-image: url('data:image/png;base64,iVB...
}`

where a leading quote is added, without a trailing one. Causes render failure, but doesn't throw an error that I can tell.

Switching to `keepBreaks: true`, or adding `rebase: false` seem to be functioning work-arounds.

OS: Windows Server 2008

Thanks!

**Comment by jakubpawlowicz** [comment]
Could you please provide more info, see https://github.com/jakubpawlowicz/clean-css/blob/master/CONTRIBUTING.md

**Comment by leggomuhgreggo** [comment]
Hey @jakubpawlowicz, I'd reviewed the contributing.md before posting--looks like I missed the version of the plugin. It's the gulp-css-minify version 1.2.1. I realize that it's a different plugin, and I'm not sure how that correlates to clean-css releases. Let me know if this issue would be better suited on the gulp-css-minify repo or if there's anything else I can do to help clarify. Thanks!

**Comment by jakubpawlowicz** [comment]
@leggomuhgreggo sorry for a bit of a lag. You can identify clean-css version with `npm ls clean-css` and that would definitely help.

**Comment by FWennerdahl** [comment]
We've run into this issue as well, also running gulp-minify-css 1.1.5 with css-clean version 3.4.6.

The issue we're having is that we run two passes of css-clean, one in our web UI project and another when building an application using said project.

Given the initial CSS (base64 shortened for clarity):
```
[icon="icon1"]:before { content:url("data:image/svg+xml;base64,PHN2ZyB4...3N2Zz4="); }

[icon="icon2"]:before { content:url("data:image/svg+xml;base64,PHN2ZyB4...3N2Zz4="); }
```

the first pass produces the following:

```
[icon=icon1]:before{content:url(data:image/svg+xml;base64,PHN2ZyB4...3N2Zz4=)}[icon=icon2]:before{content:url(data:image/svg+xml;base64,PHN2ZyB4...3N2Zz4=)}
```

Whitespace and quotes are stripped, which poses a problem when running the second pass due to the requirements mentioned at https://github.com/jakubpawlowicz/clean-css/blob/02fd73e617adf45332b6e3a5410b4285e6c99830/lib/urls/reduce.js#L43-L45.

I'm unsure why this happens, looking at `rewriteUrls` it seems like data URIs should always be output enclosed in single quotes (https://github.com/jakubpawlowicz/clean-css/blob/02fd73e617adf45332b6e3a5410b4285e6c99830/lib/urls/rewrite.js#L60-L61).

The result of running a second pass is this:

```
[icon=icon1]:before{content:url('data:image/svg+xml;base64,PHN2ZyB4...3N2Zz4=)}[icon=icon2]:before{content:url(data:image/svg+xml;base64,PHN2ZyB4...3N2Zz4=')}
```

Here single quotes are placed only at the beginning of the first data URI and at the end of the last, enclosing the second selector, effectively breaking theses styles.

Specifying the `rebase: false` option did not seem to help. When debugging I found that the rebasing happens twice when calling `minify`and that the rebase option only disabled the second occurrance.

Please let me know if you need any more information regarding this.

**Comment by jakubpawlowicz** [comment]
Thanks @FWennerdahl - this sheds some light on why it happens. I need to figure out a way to replicate it with two passes as I have had no luck so far. Could you share your clean-css related gulp setup?

**Comment by aselbie** [comment]
Adding my experience here since I believe it is related. It looks like this issue extends beyond just the data URI and is grabbing a pseudo-selector as the end point. 

- Used through ``gulp-minify-css`` version 1.2.1 with no options passed in
- Bug appeared this morning, which I assume was caused by the move from 3.4.6 to 3.4.7

# Input

```css
.icon.plus { background-image: url("data:image/svg+xml;base64,..."); }
... A bunch of lines of icons ...
ul.color-even > li:nth-child(even) { background-color: #eee; }
ul.color-odd > li:nth-child(odd) { background-color: #eee; }
```

# Output

```css
.icon.plus { background-image: url('data:image/svg+xml;base64,...) }
... A bunch of lines of icons ...
ul.color-even>li:nth-child(even), ul.color-odd>li:nth-child(odd') { background-color: #eee }
```

**Comment by bpasero** [comment]
@jakubpawlowicz also running into this as of today, this is really bad as it breaks our icons in the UI. Any ideas for a workaround?

**Comment by eternalsky** [comment]
@jakubpawlowicz also running into this problem today. I use `gulp-minify-css` version 1.2.1 & `clean-css`  version 3.4.6

## Input 
```css
.kuma-checkbox[type="radio"] + s,
.kuma-checkbox.radio-primary[type="radio"] + s {
  width: 14px;
  height: 14px;
  background-image: url("data:image/svg+xml,%3Csvg%20xmlns%3D%22http%3A%2F%2Fwww.w3.org%2F2000%2Fsvg%22%20width%3D%2215%22%20height%3D%2215%22%3E%0A%09%3Ccircle%20cx%3D%227%22%20cy%3D%227%22%20r%3D%227%22%20fill%3D%22%23ccc%22%2F%3E%0A%09%3Ccircle%20cx%3D%227%22%20cy%3D%227%22%20r%3D%226%22%20fill%3D%22%23fff%22%2F%3E%0A%3C%2Fsvg%3E%0A");
}
.kuma-checkbox[type="radio"]:checked + s {
  background-image: url("data:image/svg+xml,%3Csvg%20xmlns%3D%22http%3A%2F%2Fwww.w3.org%2F2000%2Fsvg%22%20width%3D%2215%22%20height%3D%2215%22%3E%0A%09%3Ccircle%20cx%3D%227%22%20cy%3D%227%22%20r%3D%227%22%20fill%3D%22%23f90%22%2F%3E%0A%09%3Ccircle%20cx%3D%227%22%20cy%3D%227%22%20r%3D%222.5%22%20fill%3D%22%23fff%22%2F%3E%0A%3C%2Fsvg%3E%0A");
}
```

## output 
.kuma-checkbox[type=radio]+s{width:14px;height:14px;background-image:url("data:image/svg+xml,%3Csvg%20xmlns%3D%22http%3A%2F%2Fwww.w3.org%2F2000%2Fsvg%22%20width%3D%2215%22%20height%3D%2215%22%3E%0A%09%3Ccircle%20cx%3D%227%22%20cy%3D%227%22%20r%3D%227%22%20fill%3D%22%23ccc%22%2F%3E%0A%09%3Ccircle%20cx%3D%227%22%20cy%3D%227%22%20r%3D%226%22%20fill%3D%22%23fff%22%2F%3E%0A%3C%2Fsvg%3E%0A)}.kuma-checkbox[type=radio]:checked+s{background-image:url(data:image/svg+xml,%3Csvg%20xmlns%3D%22http%3A%2F%2Fwww.w3.org%2F2000%2Fsvg%22%20width%3D%2215%22%20height%3D%2215%22%3E%0A%09%3Ccircle%20cx%3D%227%22%20cy%3D%227%22%20r%3D%227%22%20fill%3D%22%23f90%22%2F%3E%0A%09%3Ccircle%20cx%3D%227%22%20cy%3D%227%22%20r%3D%222.5%22%20fill%3D%22%23fff%22%2F%3E%0A%3C%2Fsvg%3E%0A)}

**Comment by jakubpawlowicz** [comment]
Hmmm, could be a regression, looking into it.

**Comment by jakubpawlowicz** [comment]
I've just tested it with 3.4.7 and gulp 3.9 + gulp-minify-css 1.2.1 and I still cannot reproduce this issue. Is anyone able to share some more details, like gulpfile and CSS files, with me?

**Comment by aselbie** [comment]
[app.txt](https://github.com/jakubpawlowicz/clean-css/files/32256/app.txt) is our CSS immediately before minification. We're running it through gulp-minify-css 1.2.1 without any options.

**Comment by jakubpawlowicz** [comment]
Thanks @aselbie, I can see it now but only if I pass no options to gulp plugin. I've found out passing `keepBreaks: true` as an option is a workaround. Figuring out why it happens, so stay tuned.

**Comment by eternalsky** [comment]
@jakubpawlowicz  when I change the version of clean-css to 3.4.5, this bug disappears. here is my gulpfile concerning this problem

```js
gulp.task('less', ['less_index'], function(){
    return gulp.src(['./style/kuma/src/kuma.less', './style/kuma/src/theme/*.less'])
		.pipe(inject.append('\n' + themeLessFileContent))
        .pipe(less({
            plugins: [autoprefix, LessPluginInlineUrls]
        }))
        .pipe(minifyCss())
        .pipe(gulp.dest('./theme/static/style/'));
});
```

**Comment by mareksuscak** [comment]
Started happening recently to us as well. We do use `gulp-minify-css 1.2.1`, `clean-css 3.4.7` and `gulp 3.9.0`. Any updates?

**Comment by jakubpawlowicz** [comment]
Fix is on 3.4 branch if anyone wants to test it. I'll post it with 3.4.8 as soon as tomorrow morning.

**Comment by jakubpawlowicz** [comment]
Released as 3.4.8. Thanks all for helping out!

**Comment by dentonliu** [comment]
I've got similar problem in version 3.4.12, here is sample:

**Input:**
`background-image: url('data:image/svg+xml;charset=utf-8,<svg viewBox=\'0 0 120 120\' xmlns=\'http://www.w3.org/2000/svg\' xmlns:xlink=\'http://www.w3.org/1999/xlink\'><defs><line id=\'l\' x1=\'60\' x2=\'60\' y1=\'7\' y2=\'27\' stroke=\'%236c6c6c\' stroke-width=\'11\' stroke-linecap=\'round\'/></defs><g><use xlink:href=\'%23l\' opacity=\'.27\'/><use xlink:href=\'%23l\' opacity=\'.27\' transform=\'rotate(30 60,60)\'/><use xlink:href=\'%23l\' opacity=\'.27\' transform=\'rotate(60 60,60)\'/><use xlink:href=\'%23l\' opacity=\'.27\' transform=\'rotate(90 60,60)\'/><use xlink:href=\'%23l\' opacity=\'.27\' transform=\'rotate(120 60,60)\'/><use xlink:href=\'%23l\' opacity=\'.27\' transform=\'rotate(150 60,60)\'/><use xlink:href=\'%23l\' opacity=\'.37\' transform=\'rotate(180 60,60)\'/><use xlink:href=\'%23l\' opacity=\'.46\' transform=\'rotate(210 60,60)\'/><use xlink:href=\'%23l\' opacity=\'.56\' transform=\'rotate(240 60,60)\'/><use xlink:href=\'%23l\' opacity=\'.66\' transform=\'rotate(270 60,60)\'/><use xlink:href=\'%23l\' opacity=\'.75\' transform=\'rotate(300 60,60)\'/><use xlink:href=\'%23l\' opacity=\'.85\' transform=\'rotate(330 60,60)\'/></g></svg>');`

**output:**
`background-image: url("data:image/svg+xml;charset=utf-8,<svg viewBox=\'0 0 120 120\' xmlns=\'http://www.w3.org/2000/svg\' xmlns:xlink=\'http://www.w3.org/1999/xlink\'><defs><line id=\'l\' x1=\'60\' x2=\'60\' y1=\'7\' y2=\'27\' stroke=\'%236c6c6c\' stroke-width=\'11\' stroke-linecap=\'round\'/></defs><g><use xlink:href=\'%23l\' opacity=\'.27\'/><use xlink:href=\'%23l\' opacity=\'.27\' transform=\'rotate(30 60,60)\'/><use xlink:href=\'%23l\' opacity=\'.27\' transform=\'rotate(60 60,60)\'/><use xlink:href=\'%23l\' opacity=\'.27\' transform=\'rotate(90 60,60)\'/><use xlink:href=\'%23l\' opacity=\'.27\' transform=\'rotate(120 60,60)\'/><use xlink:href=\'%23l\' opacity=\'.27\' transform=\'rotate(150 60,60)\'/><use xlink:href=\'%23l\' opacity=\'.37\' transform=\'rotate(180 60,60)\'/><use xlink:href=\'%23l\' opacity=\'.46\' transform=\'rotate(210 60,60)\'/><use xlink:href=\'%23l\' opacity=\'.56\' transform=\'rotate(240 60,60)\'/><use xlink:href=\'%23l\' opacity=\'.66\' transform=\'rotate(270 60,60)\'/><use xlink:href=\'%23l\' opacity=\'.75\' transform=\'rotate(300 60,60)\'/><use xlink:href=\'%23l\' opacity=\'.85\' transform=\'rotate(330 60,60)\'/></g></svg>');`

After minifying, the first single quote is changed to double quote, but the second is not. I have fixed it by removing the single quote. Just put comment here to help to improve code.

Sorry for my poor english
